src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,980 @@
+            TimeSynchronizationServer::Instance().SetUTCTime(commandPath.mEndpointId, utcTime, granularity, TimeSourceEnum::kAdmin))
+    {
+        commandObj->AddStatus(commandPath, Status::Success);
+    }
+    else
+    {
+        commandObj->AddClusterSpecificFailure(commandPath, to_underlying(StatusCode::kTimeNotAccepted));
+    }
+    return true;
+}
+
+bool emberAfTimeSynchronizationClusterSetTrustedTimeSourceCallback(
+    chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
+    const chip::app::Clusters::TimeSynchronization::Commands::SetTrustedTimeSource::DecodableType & commandData)
+{
+    const auto & timeSource = commandData.trustedTimeSource;
+    DataModel::Nullable<Structs::TrustedTimeSourceStruct::Type> tts;
+
+    if (!timeSource.IsNull())
+    {
+
+        Structs::TrustedTimeSourceStruct::Type ts = { commandObj->GetAccessingFabricIndex(), timeSource.Value().nodeID,
+                                                      timeSource.Value().endpoint };
+        tts.SetNonNull(ts);
+        // TODO: schedule a utctime read from this time source and emit event only on failure to get time
+        sendTimeFailureEvent(commandPath.mEndpointId);
+    }
+    else
+    {
+        tts.SetNull();
+        sendMissingTrustedTimeSourceEvent(commandPath.mEndpointId);
+    }
+
+    TimeSynchronizationServer::Instance().SetTrustedTimeSource(tts);
+    commandObj->AddStatus(commandPath, Status::Success);
+    return true;
+}
+
+bool emberAfTimeSynchronizationClusterSetTimeZoneCallback(
+    chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
+    const chip::app::Clusters::TimeSynchronization::Commands::SetTimeZone::DecodableType & commandData)
+{
+    const auto & timeZone = commandData.timeZone;
+
+    CHIP_ERROR err = TimeSynchronizationServer::Instance().SetTimeZone(timeZone);
+    if (err != CHIP_NO_ERROR)
+    {
+        if (err == CHIP_ERROR_BUFFER_TOO_SMALL)
+        {
+            commandObj->AddStatus(commandPath, Status::ResourceExhausted);
+        }
+        else if (err == CHIP_IM_GLOBAL_STATUS(InvalidCommand))
+        {
+            commandObj->AddStatus(commandPath, Status::InvalidCommand);
+        }
+        else
+        {
+            commandObj->AddStatus(commandPath, Status::ConstraintError);
+        }
+        return true;
+    }
+
+    if (to_underlying(TimeSynchronizationServer::Instance().GetEventFlag()) & to_underlying(TimeSyncEventFlag::kTimeZoneStatus))
+    {
+        TimeSynchronizationServer::Instance().ClearEventFlag(TimeSyncEventFlag::kTimeZoneStatus);
+        sendTimeZoneStatusEvent(commandPath.mEndpointId);
+    }
+    GetDelegate()->HandleTimeZoneChanged(TimeSynchronizationServer::Instance().GetTimeZone());
+
+    TimeZoneDatabaseEnum tzDb;
+    TimeZoneDatabase::Get(commandPath.mEndpointId, &tzDb);
+    Commands::SetTimeZoneResponse::Type response;
+    TimeSynchronizationServer::Instance().GetUpdatedTimeZoneState();
```

Comment:
See, this is why I was suggesting in my previous review we rename that function to UpdateTimeZoneState() (and similar for DST offset state).  It makes no sense to call a Get... thing and ignore the return value!

Followup is ok, but please don't lose track of this.

Reply:
Tracked in https://github.com/project-chip/connectedhomeip/issues/27575, resolving for now.